ray-on-gke/kuberay-tpu-webhook/main.go

```diff
@@ -88,74 +93,192 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
 	return &rayCluster, nil
 }
 
-func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec) string {
-	numWorkers := int(*workerGroupSpec.Replicas)
+func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec) (string, error) {
+	replicas := workerGroupSpec.Replicas
+	if replicas == nil {
+		return "", errors.New("workerGroupSpec replicas not set")
+	}
+	numWorkers := int(*replicas)
 	workerGroupName := workerGroupSpec.GroupName
 	serviceName := "tpu-worker-group-svc" // TODO: Kuberay should set DNS name for workergroup
 	hostNames := make([]string, numWorkers)
 	for j := 0; j < numWorkers; j++ {
 		hostNames[j] = fmt.Sprintf("%s-%d.%s", workerGroupName, j, serviceName)
 	}
-	return strings.Join(hostNames, ",")
+	return strings.Join(hostNames, ","), nil
 }
 
 func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, workerGroupIndex int, patches *[]patch) {
-	for j := 0; j < len(workerGroupSpec.Template.Spec.Containers); j++ {
-		container := workerGroupSpec.Template.Spec.Containers[j]
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		klog.Fatalf("Container path not specified")
+	}
+	for j := 0; j < len(containers); j++ {
+		container := containers[j]
 		if containerRequestingTPUs(container) {
-			hostNamesPatch := patch {
-				"op": "add",
-			}
-			path := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
+			subdomainPatch, hostNamesPatch := patch{"op": "add",}, patch{"op": "add",}
+			subdomainPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/subdomain", workerGroupIndex)
+			envPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
 			tpuWorkerHostNames := corev1.EnvVar{
 				Name:  "TPU_WORKER_HOSTNAMES",
 				Value: hostNames,
 			}
+			subdomainPatch["path"] = subdomainPath
+			subdomainPatch["value"] = "tpu-worker-group-svc"
 			if len(container.Env) == 0 {
-				hostNamesPatch["path"] = path
+				hostNamesPatch["path"] = envPath
 				hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
 			} else {
-				hostNamesPatch["path"] = fmt.Sprintf("%s/-", path)
+				hostNamesPatch["path"] = fmt.Sprintf("%s/-", envPath)
 				hostNamesPatch["value"] = tpuWorkerHostNames
 			}
-			*patches = append(*patches, hostNamesPatch)
+			*patches = append(*patches, subdomainPatch, hostNamesPatch)
+		}
+	}
+}
+
+func checkWorkersMatchTopology(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	replicas := workerGroupSpec.Replicas
+	if replicas == nil {
+		return false, errors.New("workerGroupSpec replicas not set")
+	}
+	numWorkers := int(*replicas)
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		return false, errors.New("Container path not specified")
+	}
+	if containerRequestingTPUs(containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		if topology == "" {
+			klog.Error("TPU topology not specified")
```

Comment:
Return err here?

Reply:
Returning an error here actually breaks the webhook - I believe it's because the tpu-topology may not be set yet and we want the webhook to keep running rather than exit out.